File: leetcode/243-shortest-word-distance/main.py
"""
  questions to ask:
  - will word1 == word2?
  - will either word1, word2 be in the list always? 
"""

"""
    naive solution: brute force O(n^2)
"""


class Solution(object):
    def shortestDistance(self, words, word1, word2):
        """
        :type words: List[str]
        :type word1: str
        :type word2: str
        :rtype: int

        1st approach: 2pointers

        Time    O(n)
        Space   O(n)
        52 ms, faster than 27.57%
        """
        p1 = -1
        p2 = -1
        res = len(words)
        for i in range(len(words)):
            word = words[i]
            if word == word1:
                p1 = i
            if word == word2:
                p2 = i
            if p1 != -1 and p2 != -1:
                res = min(res, abs(p1-p2))
        return res


a = ["practice", "makes", "perfect", "coding", "makes"]
b = "makes"
c = "coding"
print(Solution().shortestDistance(a, b, c))

print("-----")


class Solution(object):
    def shortestDistance(self, words, word1, word2):
        """
        1st approach: binary search the nearest target index

        Time    O(mlogn) m: word1, n: word2
        Space   O(m+n)
        12 ms, faster than 7.14%
        """
        A, B = [], []
        for i in range(len(words)):
            w = words[i]
            if w == word1:
                A.append(i)
            elif w == word2:
                B.append(i)
        res = 2**32
        for x in A:
            j = self.bsearch(B, x)
            y = B[j]
            res = min(res, abs(x-y))
        return res

    def bsearch(self, nums, target):
        left = 0
        right = len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if target < nums[mid]:
                right = mid - 1
            elif target > nums[mid]:
                left = mid + 1
            else:
                return mid
        if right < 0:
            return 0
        if left > len(nums)-1:
            return len(nums)-1
        if abs(nums[left] - target) < abs(nums[right] - target):
            return left
        return right

File: leetcode/243-shortest-word-distance/main.go
package main

import "fmt"

/*
	questions to ask:
  - will word1 == word2?
  - will either word1, word2 be in the list always?
*/

/*
	1st approach: 2 pointers

	Time    O(n)
	Space   O(1)
	4ms beats 100%
*/
func shortestDistance(words []string, word1 string, word2 string) int {
	p1 := -1
	p2 := -1
	res := len(words)
	for i := 0; i < len(words); i++ {
		word := words[i]
		if word == word1 {
			p1 = i
		}
		if word == word2 {
			p2 = i
		}
		diff := abs(p1 - p2)
		if p1 != -1 && p2 != -1 && diff < res {
			res = diff
		}
	}
	return res
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

/*
	1st approach: binary search the nearest target index

	Time    O(mlogn) m: word1, n: word2
	Space   O(m+n)
	12 ms, faster than 7.14%
*/
func shortestDistance1(words []string, word1 string, word2 string) int {
	m := make(map[string][]int)
	for i := 0; i < len(words); i++ {
		word := words[i]
		if _, x := m[word]; !x {
			m[word] = []int{i}
		} else {
			m[word] = append(m[word], i)
		}
	}
	arr1 := m[word1]
	arr2 := m[word2]
	res := len(words)
	// binary search
	for _, targetIdx := range arr1 {
		left := 0
		right := len(arr2) - 1
		for left <= right {
			mid := (left + right) / 2
			// compare with the result
			diff := abs(targetIdx - arr2[mid])
			if diff < res {
				res = diff
			}
			// binary search range narrowing
			if arr2[mid] < targetIdx {
				left = mid + 1
			} else {
				right = mid - 1
			}
		}
	}
	return res
}

func main() {
	a := []string{"practice", "makes", "perfect", "coding", "makes"}
	fmt.Println(shortestDistance1(a, "coding", "practice"))

	a = []string{"practice", "makes", "perfect", "coding", "makes"}
	fmt.Println(shortestDistance1(a, "coding", "makes"))
}
